### PR TITLE
network: ndisc - Also fill router preference

### PR DESCRIPTION
**Pull request by ssahani**
May fix https://github.com/systemd/systemd/issues/28426

### CODE REVIEW

File: src/network/networkd-ndisc.c [hunk 1]
```diff
@@ -423,9 +423,9 @@ static int ndisc_router_process_autonomous_prefix(Link *link, sd_ndisc_router *r
 static int ndisc_router_process_onlink_prefix(Link *link, sd_ndisc_router *rt) {
         _cleanup_(route_freep) Route *route = NULL;
         unsigned prefixlen, preference;
-        usec_t timestamp_usec;
-        uint32_t lifetime_sec;
         struct in6_addr prefix;
+        uint32_t lifetime_sec;
+        usec_t timestamp_usec;
```
Comment (yuwata) [inline src/network/networkd-ndisc.c]: Please drop this unnecessary change.

File: src/network/networkd-ndisc.c [hunk 2]
```diff
@@ -1018,6 +1018,51 @@ static int ndisc_drop_outdated(Link *link, usec_t timestamp_usec) {
         return r;
 }
 
+static int ndisc_drop_lifetime_zero_router(Link *link, sd_ndisc_router *rt) {
+        struct in6_addr router, prefix;
```
Comment (yuwata) [inline src/network/networkd-ndisc.c]: The second commit should be redundant. Already addressed by af2aea8bb64b0dc42ecbe5549216eb567681a803.

File: src/network/networkd-ndisc.c [hunk 3]
```diff
@@ -451,13 +451,18 @@ static int ndisc_router_process_onlink_prefix(Link *link, sd_ndisc_router *rt) {
         if (r < 0)
                 return log_link_warning_errno(link, r, "Failed to get prefix length: %m");
 
+        r = sd_ndisc_router_get_preference(rt, &preference);
```
Comment (yuwata) [inline src/network/networkd-ndisc.c:481]: Are there any relevant statement in an RFC about that the 'main' preference should/may be used for routes derived by the on-link prefix ?

File: src/network/networkd-ndisc.c (hunk 2 above)
Comment (ssahani) [reply src/network/networkd-ndisc.c]: it did not work. The routes were still there.

File: src/network/networkd-ndisc.c (hunk 3 above)
Comment (ssahani) [reply src/network/networkd-ndisc.c:481]: I did not find any. If omitted then pref always be  medium even if it's from high/low pref. Hence it's not behaving as is supposed to do. https://github.com/systemd/systemd/issues/28439 and https://github.com/systemd/systemd/issues/28426. 

Here is a verified test https://github.com/systemd/systemd/issues/28426#issuecomment-1653398818

**Comment by ssahani** [comment]
Updated thanks for the review !

File: src/network/networkd-ndisc.c [hunk 4]
```diff
@@ -1013,6 +1018,51 @@ static int ndisc_drop_outdated(Link *link, usec_t timestamp_usec) {
...
+                        if (r < 0)
+                                log_link_warning_errno(link, r, "Failed to remove outdated SLAAC route, ignoring: %m");
+                }
+
+                if (route->family == AF_INET6 && memcmp(&route->dst.in6, &prefix, sizeof(route->dst.in6) == 0) &&
+                    route->dst_prefixlen == prefixlen && lifetime_sec == 0)  {
+                        r = route_remove_and_drop(route);
+                        if (r < 0)
+                                log_link_warning_errno(link, r, "Failed to remove outdated SLAAC route, ignoring: %m");
+                }
```
Comment (yuwata) [inline src/network/networkd-ndisc.c]: Such comparison is not necessary.
Please check `route->provider` and the router address.

File: src/network/networkd-ndisc.c [hunk 5]
```diff
@@ -1013,6 +1018,51 @@ static int ndisc_drop_outdated(Link *link, usec_t timestamp_usec) {
         return r;
 }
 
+static int ndisc_drop_lifetime_zero_router(Link *link, sd_ndisc_router *rt) {
```
Comment (yuwata) [inline src/network/networkd-ndisc.c]: Please rename `ndisc_process_zero_lifetime()` or so.
Then, also remove SLAAC, DNS, Domains, Captive portals.

File: src/network/networkd-ndisc.c [hunk 6]
```diff
@@ -1159,15 +1210,31 @@ static int ndisc_router_handler(Link *link, sd_ndisc_router *rt) {
...
         if (r < 0)
                 return r;
 
-        r = ndisc_router_process_default(link, rt);
+        /* https://datatracker.ietf.org/doc/html/rfc4861
+         * Router Lifetime: A Lifetime of 0 indicates that the router is not a default router
+         * and SHOULD NOT appear on the default router list.
+         */
+        if (router_lifetime_sec == 0) {
+                log_link_debug(link, "Received RA with lifetime = 0, dropping route.");
```
Comment (yuwata) [inline src/network/networkd-ndisc.c]: Then, `dropping route.` -> `dropping configurations.`, as all configurations based on the previous RA from the router will be dropped.

File: src/network/networkd-ndisc.c [hunk 7]
```diff
@@ -1159,15 +1210,31 @@ static int ndisc_router_handler(Link *link, sd_ndisc_router *rt) {
...
+
+                r = ndisc_drop_lifetime_zero_router(link, rt);
+                if (r < 0)
+                        log_link_warning_errno(link, r, "Failed to remove SLAAC route having lifetime = 0, ignoring: %m");
+        } else {
```
Comment (yuwata) [inline src/network/networkd-ndisc.c]: When lifetime is zero, it is not necessary to call anything below. So, return earlier.

File: src/network/networkd-ndisc.c (hunk 3 above)
Comment (yuwata) [inline src/network/networkd-ndisc.c:481]: Could you comment that Prefix Information option does not have preference, hence we use the 'main' preference here ?

**Review by yuwata — CHANGES_REQUESTED** [review]

**Comment by ssahani** [comment]
Updated. Thanks for the review PTAL

File: src/network/networkd-ndisc.c [hunk 8]
```diff
@@ -1175,15 +1226,33 @@ static int ndisc_router_handler(Link *link, sd_ndisc_router *rt) {
...
-        r = ndisc_drop_outdated(link, timestamp_usec);
+        r = sd_ndisc_router_get_lifetime(rt, &router_lifetime_sec);
         if (r < 0)
-                return r;
+                return log_link_warning_errno(link, r, "Failed to get lifetime of RA message: %m");
 
-        r = ndisc_start_dhcp6_client(link, rt);
+        r = ndisc_drop_outdated(link, timestamp_usec);
         if (r < 0)
                 return r;
```
Comment (yuwata) [inline src/network/networkd-ndisc.c]: Please do not mix the order.
1. check router address and filter the message if necessary,
2. check lifetime and drop previous assignments if lifetime is zero,
3. check timestamp and drop outdated assignments if necessary.

```c
r = sd_ndisc_router_get_lifetime(rt, &router_lifetime_sec);
if (r < 0)
        return log_link_warning_errno(link, r, ...);

if (router_lifetime_sec == 0) {
        r = ndisc_process_zero_lifetime(...);
        if (r < 0)
                log_link_warning_errno(...);
        return 0;
}

r = sd_ndisc_router_get_timestamp(rt, CLOCK_BOOTTIME, &timestamp_usec);
if (r == -ENODATA)
        ...
if (r < 0)
        ...

r = ndisc_drop_outdated(link, timestamp_usec);
if (r < 0)
        return r;

r = ndisc_start_dhcp6_client(...);
if (r < 0)
        return r;
```

File: src/network/networkd-ndisc.c [hunk 9]
```diff
@@ -1175,15 +1226,33 @@ static int ndisc_router_handler(Link *link, sd_ndisc_router *rt) {
...
 
+        /* https://datatracker.ietf.org/doc/html/rfc4861
+         * Router Lifetime: A Lifetime of 0 indicates that the router is not a default router
+         * and SHOULD NOT appear on the default router list.
+         */
+        if (router_lifetime_sec == 0) {
+                log_link_debug(link, "Received RA with lifetime = 0, dropping configurations.");
```
Comment (yuwata) [inline src/network/networkd-ndisc.c:1225]: Not a matter, but I'd like to move the log message into `ndisc_process_zero_lifetime()`.

File: src/network/networkd-ndisc.c [hunk 10]
```diff
@@ -1029,6 +1029,56 @@ static int ndisc_drop_outdated(Link *link, usec_t timestamp_usec) {
...
+        NDiscRDNSS *rdnss;
+        Address *address;
+        Route *route;
+        int r, k;
+
+        assert(link);
+
+        SET_FOREACH(route, link->routes) {
+                if (route->source != NETWORK_CONFIG_SOURCE_NDISC)
+                        continue;
```
Comment (yuwata) [inline src/network/networkd-ndisc.c]: Please also check if `route->provider` matches the router address.
There may be multiple routers in the network.

File: src/network/networkd-ndisc.c [hunk 11]
```diff
@@ -1029,6 +1029,56 @@ static int ndisc_drop_outdated(Link *link, usec_t timestamp_usec) {
...
+
+                r = route_remove_and_drop(route);
+                if (r < 0)
+                        log_link_warning_errno(link, r, "Failed to remove outdated SLAAC route, ignoring: %m");
+        }
+
+        SET_FOREACH(address, link->addresses) {
+                if (address->source != NETWORK_CONFIG_SOURCE_NDISC)
+                        continue;
```
Comment (yuwata) [inline src/network/networkd-ndisc.c]: Also here.

File: src/network/networkd-ndisc.c [hunk 12]
```diff
@@ -1029,6 +1029,56 @@ static int ndisc_drop_outdated(Link *link, usec_t timestamp_usec) {
...
+
+                k = address_remove_and_drop(address);
+                if (k < 0)
+                        r = log_link_warning_errno(link, k, "Failed to remove outdated SLAAC address, ignoring: %m");
+        }
+
+        SET_FOREACH(rdnss, link->ndisc_rdnss) {
```
Comment (yuwata) [inline src/network/networkd-ndisc.c]: Here also and below. please check if `rdnss->router` matches the router address.

**Review by yuwata — CHANGES_REQUESTED** [review]
Sorry for late review. We recently received several regression reports in the last release, and I was busy to address the issues.

The first commit LGTM.

**Comment by ssahani** [comment]
updated. Thanks for the review

File: src/network/networkd-ndisc.c [hunk 13]
```diff
@@ -1029,6 +1029,69 @@ static int ndisc_drop_outdated(Link *link, usec_t timestamp_usec) {
         return r;
 }
 
+static int ndisc_process_zero_lifetime(Link *link, sd_ndisc_router *rt) {
```
Comment (keszybz) [inline src/network/networkd-ndisc.c]: This functions seems to be almost a copy of `ndisc_drop_outdated()`. Instead of duplicating this logic here, just call `ndisc_drop_outdated(timestamp_usec=INFINITY)`. Maybe some additional conditionalization of `ndisc_drop_outdated()` will be needed… But it's certainly better than having two functions doing almost the same thing.

Comment (ssahani) [reply src/network/networkd-ndisc.c]: Thanks for taking a look. We need the ` sd_ndisc_router` object. There could be a multiple routers in the network with diff precedence.  `ndisc_drop_outdated` does not have this object.  We are comparing with the `provider`  i.e. from which source this resource got configured. and dropping.

File: src/network/networkd-ndisc.c [hunk 14]
```diff
@@ -1029,6 +1029,69 @@ static int ndisc_drop_outdated(Link *link, usec_t timestamp_usec) {
...
+        assert(link);
+
+        log_link_debug(link, "Received RA with lifetime = 0, dropping configurations.");
+
+        r = sd_ndisc_router_get_address(rt, &router);
+        if (r < 0)
+                return log_link_warning_errno(link, r, "Failed to get gateway address from RA: %m");
+
+        SET_FOREACH(route, link->routes) {
+                if (route->source != NETWORK_CONFIG_SOURCE_NDISC && memcmp(&route->provider.in6, &router, sizeof(route->provider.in6)) != 0)
```
Comment (yuwata) [inline src/network/networkd-ndisc.c]: Please use `in6_addr_equal()`, instead of `memcmp()`. Also, `&&` -> `||`, I think it is better to split the condition:
```c
if (route->source != NETWORK_CONFIG_SOURCE_NDISC)
        continue;

if (!in6_addr_equal(&route->provider.in6, &router))
        continue;
```

File: src/network/networkd-ndisc.c [hunk 15]
```diff
@@ -1029,6 +1029,69 @@ static int ndisc_drop_outdated(Link *link, usec_t timestamp_usec) {
...
+                        continue;
+
+                r = route_remove_and_drop(route);
+                if (r < 0)
+                        log_link_warning_errno(link, r, "Failed to remove outdated SLAAC route, ignoring: %m");
+        }
+
+        SET_FOREACH(address, link->addresses) {
+                if (address->source != NETWORK_CONFIG_SOURCE_NDISC && memcmp(&address->provider.in6, &router, sizeof(address->provider.in6)) != 0)
```
Comment (yuwata) [inline src/network/networkd-ndisc.c]: Also here, and below.

File: src/network/networkd-ndisc.c [hunk 16]
```diff
@@ -1175,15 +1239,31 @@ static int ndisc_router_handler(Link *link, sd_ndisc_router *rt) {
...
 
-        r = ndisc_start_dhcp6_client(link, rt);
+        /* https://datatracker.ietf.org/doc/html/rfc4861
+         * Router Lifetime: A Lifetime of 0 indicates that the router is not a default router
+         * and SHOULD NOT appear on the default router list.
+         */
+        if (router_lifetime_sec == 0) {
+                r = ndisc_process_zero_lifetime(link, rt);
+                if (r < 0)
+                        log_link_warning_errno(link, r, "Failed to remove SLAAC route having lifetime = 0, ignoring: %m");
```
Comment (yuwata) [inline src/network/networkd-ndisc.c]: The function drops not only route. Maybe, `"Failed to process RA with zero lifetime, ignoring: %m"` is enough.

File: src/network/networkd-ndisc.c [hunk 17]
```diff
@@ -1175,15 +1239,31 @@ static int ndisc_router_handler(Link *link, sd_ndisc_router *rt) {
...
+
+        r = ndisc_drop_outdated(link, timestamp_usec);
         if (r < 0)
                 return r;
 
         r = ndisc_router_process_default(link, rt);
+        if (r < 0)
+               return r;
+
+        r = ndisc_start_dhcp6_client(link, rt);
```
Comment (yuwata) [inline src/network/networkd-ndisc.c]: Hmm. Please do not unnecessarily change the order of the functions.
1. drop the outdated configurations,
2. start DHCPv6 client,
3. process default.

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/network/networkd-ndisc.c (hunk 13 above)
Comment (yuwata) [reply src/network/networkd-ndisc.c]: Yeah, @keszybz's suggestion sounds nice.
Please merge the two functions like this.
```c
static int ndisc_drop_outdated(Link *link, usec_t timestamp_usec, struct in6_addr *router) {
        ...
        SET_FOREACH(route, link->routes) {
                if (route->source != NETWORK_CONFIG_SOURCE_NDISC)
                        continue;

                if (route->lifetime_usec >= timestamp_usec)
                        continue; /* the route is still valid */

                if (router && !in6_addr_equal(&route->provider.in6, router))
                        continue;

                k = route_remove_and_drop(route);
                if (k < 0)
                        r = log_link_warning_errno(link, k, "Failed to remove outdated SLAAC route, ignoring: %m");
        }
        ...
}
```

**Comment by ssahani** [comment]
Updated . Thanks for the review !

File: src/network/networkd-ndisc.c [hunk 18]
```diff
@@ -1192,12 +1208,26 @@ static int ndisc_router_handler(Link *link, sd_ndisc_router *rt) {
...
+         /* https://datatracker.ietf.org/doc/html/rfc4861
+          * Router Lifetime: A Lifetime of 0 indicates that the router is not a default router
+          * and SHOULD NOT appear on the default router list.
+          */
+        if (router_lifetime_sec == 0) {
+                log_link_debug(link, "Received RA with lifetime = 0, dropping configurations.");
+
+                r = ndisc_drop_outdated(link, USEC_INFINITY, &router);
+                if (r < 0)
+                        log_link_warning_errno(link, r, "Failed to process RA with zero lifetime, ignoring:: %m");
```
Comment (yuwata) [inline src/network/networkd-ndisc.c]: Doubled `:` in the log message.

Also, `return 0;` in this `if` block is mistakenly dropped:
```c
if (router_lifetime_sec == 0) {
        log_link_debug(...);

        r = ndisc_drop_outdated(...);
        if (r < 0)
                log_link_warning_errno(...);

        return 0;
}
```

**Review by yuwata — CHANGES_REQUESTED** [review]
Looks mostly OK. But one more round please.

File: src/network/networkd-ndisc.c (hunk 18 above)
Comment (ssahani) [reply src/network/networkd-ndisc.c]: Opps sry missed it

File: src/network/networkd-ndisc.c [hunk 19]
```diff
@@ -1192,12 +1208,26 @@ static int ndisc_router_handler(Link *link, sd_ndisc_router *rt) {
...
+         /* https://datatracker.ietf.org/doc/html/rfc4861
+          * Router Lifetime: A Lifetime of 0 indicates that the router is not a default router
+          * and SHOULD NOT appear on the default router list.
+          */
+        if (router_lifetime_sec == 0) {
+                log_link_debug(link, "Received RA with lifetime = 0, dropping configurations.");
+
+                r = ndisc_drop_outdated(link, USEC_INFINITY, &router);
+                if (r < 0)
+                        log_link_warning_errno(link, r, "Failed to process RA with zero lifetime, ignoring: %m");
```
Comment (yuwata) [inline src/network/networkd-ndisc.c:1229]: Still `return 0;` is missing...

**Review by yuwata — CHANGES_REQUESTED** [review]

**Review by yuwata — APPROVED** [review]
LGTM.

File: src/network/networkd-ndisc.c [hunk 20]
```diff
@@ -972,7 +978,7 @@ static int ndisc_router_process_options(Link *link, sd_ndisc_router *rt) {
         }
 }
 
-static int ndisc_drop_outdated(Link *link, usec_t timestamp_usec) {
+static int ndisc_drop_outdated(Link *link, usec_t timestamp_usec, struct in6_addr *router) {
```
Comment (yuwata) [inline src/network/networkd-ndisc.c]: Sorry, that's my fault, but please constify `router`.
